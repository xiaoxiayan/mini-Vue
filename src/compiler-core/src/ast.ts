export enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPUND_EXPRESSION
}