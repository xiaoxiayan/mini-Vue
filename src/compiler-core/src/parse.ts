import { NodeTypes } from "./ast"
const enum TagType {
  Start,
  End
}

export function baseParse(content: string) {
  const context = createParseContext(content)
  return createRoot(parseChildren(context, []))
}
// 抽离对象

function createParseContext(content) {
  return {
    sourse: content
  }
}

function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT
  }
}
// ancestors 祖先，收集 标签
function parseChildren(context, ancestors) {
  const nodes: any = []
  while (!isEnd(context, ancestors)) {
  let node
  const s = context.sourse
  // 需要循环取调用解析，直到没有值。
    if (s.startsWith('{{')) {
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors)
      }
    }
    if (!node) {
      node = parseText(context)
    }
    nodes.push(node)
  }
  return nodes
}

function parseInterpolation(context) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'
  //  {{message}}
  const closeIndex = context.sourse.indexOf(closeDelimiter, openDelimiter.length)
  // 向前推进
  advanceBy(context, openDelimiter.length)
  // 计算出中间的长度，然后截取
  const rawContentLength = closeIndex - openDelimiter.length
  const rawcontent = parseTextData(context, rawContentLength)
  const content = rawcontent.trim()
  // 然后继续推进
  console.log(context.sourse, rawContentLength + closeDelimiter.length, '长度')
  advanceBy(context, closeDelimiter.length)
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content
    }
  }
}

function advanceBy(context: any, length: number) {
  context.sourse = context.sourse.slice(length)
}

function parseElement(context, ancestors) {
  const element: any = parseTag(context, TagType.Start)
  // 收集出现过的标签
  ancestors.push(element)
  element.children = parseChildren(context, ancestors);
  // 弹出推进过的 标签
  ancestors.pop()
  // 需要判断一下， ancestors 和 当前的 标签相同才 推进。
  if(startsWithEndTagOpen (context.sourse, element.tag)) {
    parseTag(context, TagType.End)
  } else {
      throw new Error(`缺少结束标签:${element.tag}`)
  }
  return element
}
function startsWithEndTagOpen (sourse, tag) {
  return sourse.startsWith('</') && sourse.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
}
function parseTag (context: any, type: TagType) {
  // 1.解析。tag 。正则
  const match: any = /^<\/?([a-z]*)/i.exec(context.sourse)
  const tag = match[1]
  // 2.推进删除代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)
  if (type === TagType.End) return
  return {
    type: NodeTypes.ELEMENT,
    tag: tag
  }
}

function parseText(context: any): any {
  // 推进， 删除
  let endIndex = context.sourse.length
  let endTokens = ['<', '{{']
  for(let i = 0; i < endTokens.length; i++) {
    const index = context.sourse.indexOf(endTokens[i])
    if (index !== -1 && index < endIndex) {
      endIndex = index
    }
  }
  const content = parseTextData(context, endIndex)
  return {
    type: NodeTypes.TEXT,
    content
  }
}
function parseTextData(context, length) {
  const content = context.sourse.slice(0, length)
  advanceBy(context, length)
  return content
}
function isEnd(context: any, ancestors) {
  const s = context.sourse
  // 2. 遇到结束标签的时候 , 循环收集过的 ancestors， 如果有相同的就跳出循环，防止死循环

    if ( s.startsWith('</')) {
      for(let i = ancestors.length - 1 ; i >= 0; i-- ) {
        const tag = ancestors[i].tag
        if (startsWithEndTagOpen(s, tag))
           return true
        }
  }
  // 1. sourse 有值的时候
  return !s
}

