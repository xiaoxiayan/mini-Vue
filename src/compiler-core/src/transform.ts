import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  // 1. 遍历- 深度优先搜索
  traverseNode(root, context)

  // 创建 codegen 能直接使用的 codegenNode
  createCodegenNode(root)

  // 在根结点的时候 挂在一个 helper
  root.helpers = [...context.helpers.keys()]
}


function traverseNode(node: any, context) {
  // 取出插入的 方法， 调用。
  const nodeTransforms = context.nodeTransforms
  const exitFns: any = []

  for(let i = 0; i< nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    const onExit = transform(node, context)
    if(onExit) exitFns.push(onExit)
  }
  // 如果节点是 interpolation 插入节点，我们需要挂一个 helpers 的一个相关函数给 codegen，
  // 给 context去设置一个  helpers
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChilren(node, context)
      break;
    default:
      break;
  }
  let i = exitFns.length
  while(i--) {
    exitFns[i]()
  }

}

function traverseChilren(node: any, context: any) {
  const children = node.children
  for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(children[i], context)
  }

}

function createTransformContext(root: any, options: any) {
  // 生成一个对象，包含插入方法。
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    }
  }
  return context
}

function createCodegenNode(root: any) {
  // 如果类型是 element 的， 我们直接使用 element上面的 codegenNode
  const child = root.children[0]
  if(child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode
  } else {
      root.codegenNode = root.children[0]
  }
}

