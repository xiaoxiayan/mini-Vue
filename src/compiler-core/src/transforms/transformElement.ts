import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      context.helper(CREATE_ELEMENT_VNODE)
      // 中间处理层
      // tag
      const vnodeTag = node.tag
      // props
      let vnodeProps;
      // children
      const children = node.children
      //  不是很合理的地方
      let vnodeChildren = children[0]
      const vnodeElement = {
        type: NodeTypes.ELEMENT,
        tag: vnodeTag,
        props: vnodeProps,
        children: vnodeChildren
      }
      node.codegenNode = vnodeElement
    }
  }

}