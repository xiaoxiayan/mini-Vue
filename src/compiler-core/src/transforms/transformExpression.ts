import { NodeTypes } from "../ast";

export function transformExpression (node) {
    if (node.type === NodeTypes.INTERPOLATION) {
       node.content = proessExpression(node.content)
    }
}
function proessExpression(node: any) {
  node.content = `_ctx.${node.content}`
  return node
}

