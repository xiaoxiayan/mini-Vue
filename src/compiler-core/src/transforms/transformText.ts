import { NodeTypes } from "../ast";
import { isText } from "../utils";


export function transformText(node) {

  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const { children } = node
      let currentContainer
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isText(child)) {
          // 如果连续的节点是 text， interpolation
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j];
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPUND_EXPRESSION,
                  children: [child]
                }
              }
              currentContainer.children.push(' + ')
              currentContainer.children.push(next)
              children.splice(j, 1)
              // 添加以后删除，然后i 会变， 所以 j要 --
              j--
            } else {
              currentContainer = undefined
              break
            }
          }
        }
      }
    }
  }
}
