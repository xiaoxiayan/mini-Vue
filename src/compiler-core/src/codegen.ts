import { isString } from "../../shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"
import { transform } from "./transform"

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  genFuncitonPreamble(ast, context)
  let functionName = 'render'
  let args = ['_ctx', '_cache']
  const signature = args.join(',')
  push(`function ${functionName}(${signature}){`)
  push('return ')
  // 通过 ast 树去获取的 return 内容
  genNode(ast.codegenNode, context)
  push('}')
  return {
    code: context.code
  }
}

function genFuncitonPreamble(ast, context) {
  const { push } = context
  const vueBinging = 'vue'
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`
  if (ast.helpers.length > 0) {
    push(`const {${ast.helpers.map(aliasHelper).join(', ')}} = ${vueBinging}`)
  }
  push('\n')
  push('return ')
}

function genNode(node: any, context: any) {
  // 区分node类型，去return 不同的内容
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break;
    case NodeTypes.INTERPOLATION:
      genInterPolation(node, context)
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      getExperssion(node, context)
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break;
    case NodeTypes.COMPUND_EXPRESSION:
      genCompoundExpression(node, context)
      break;
    default:
      break;
  }
}


function genText(node: any, context: any) {
  const { push } = context
  push(`"${node.content}"`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(sourse) {
      context.code += sourse
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context
}

function genInterPolation(node: any, context: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')

}

function getExperssion(node: any, context: any) {
  // 专门处理表达式的类型
  const { push } = context
  push(`${node.content}`)
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, children } = node
  console.log('genElement---', children);
  push(`${helper(CREATE_ELEMENT_VNODE)}("${tag}", null, `)
  // 可以先去循环这个children
  genNode(children, context)
  push(')')
  // 创建新的节点类型， 判断是不是连续的 text, 插值
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context
  const children = node.children
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

