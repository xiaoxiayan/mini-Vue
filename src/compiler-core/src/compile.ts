import { generate } from "./codegen"
import { baseParse } from "./parse"
import { transform } from "./transform"
import { transformElement } from "./transforms/transformElement"
import { transformExpression } from "./transforms/transformExpression"
import { transformText } from "./transforms/transformText"

export function baseCompile(template) {
  const ast:any = baseParse(template)
  // 把ast 树 梳理成 render函数字符串
  transform(ast, {
    nodeTransforms: [ transformExpression , transformElement, transformText ]
  })

  return generate(ast)

}