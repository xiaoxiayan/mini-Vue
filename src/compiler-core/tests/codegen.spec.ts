import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transformText } from "../src/transforms/transformText"

describe('codegen', () => {
  it('string', () => {
       const ast = baseParse('hi')
       // 把ast 树 梳理成 render函数字符串
       transform(ast)
       const {code} = generate(ast)
       // 快照
       expect(code).toMatchSnapshot()
  })
  it('interPolation', () => {
    const ast = baseParse('{{message}}')
    // 把ast 树 梳理成 render函数字符串
    transform(ast, {
      nodeTransforms: [transformExpression]
    })
    const {code} = generate(ast)
    // 快照
    expect(code).toMatchSnapshot()
  })
  it('element', () => {
    const ast:any = baseParse('<div>h1,{{message}}</div>')
    // 把ast 树 梳理成 render函数字符串
    transform(ast, {
      nodeTransforms: [ transformExpression , transformElement, transformText ]
    })
    console.log('ast==', ast, ast.codegenNode.children)
    const {code} = generate(ast)
    // 快照
    expect(code).toMatchSnapshot()
  })
})