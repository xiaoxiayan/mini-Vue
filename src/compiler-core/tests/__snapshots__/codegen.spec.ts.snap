// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const {toDisplayString:_toDisplayString, createElementVnode:_createElementVnode} = vue
return function render(_ctx,_cache){return _createElementVnode(\\"div\\", null, \\"h1,\\" + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interPolation 1`] = `
"const {toDisplayString:_toDisplayString} = vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx,_cache){return \\"hi\\"}"
`;
